### PR TITLE
Compile JIT sources with /W4

### PR DESCRIPTION
**Pull request by mikedn**
This enables VC++'s `/W4` for JIT sources. In addition to helping catching potential coding issues this also matches what the internal .NET Framework builds seem to use so it prevents code going in that compiles fine in CoreCLR and fails in the internal build.

Apart from `/W4` it looks like the internal build uses similar compilation options, except for some unknown `guard` options.

### CODE REVIEW

**Comment by benaadams** [comment]
If you are feeling spicy `/d2FH4` too? https://github.com/dotnet/coreclr/issues/23167

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

File: src/jit/gentree.cpp [hunk 1]
```diff
@@ -18036,8 +18036,9 @@ bool GenTreeHWIntrinsic::OperIsMemoryLoadOrStore()
 {
 #ifdef _TARGET_XARCH_
     return OperIsMemoryLoad() || OperIsMemoryStore();
-#endif // _TARGET_XARCH_
+#else  // _TARGET_XARCH_
```
Comment (sandreenko) [inline src/jit/gentree.cpp]: The #else comment should be !_TARGET_XARCH_ (and endif as well).

File: src/jit/hashbv.cpp [hunk 2]
```diff
@@ -793,15 +793,16 @@ void hashBv::setBit(indexType index)
...
     indexType baseIndex = index & ~(BITS_PER_NODE - 1);
     indexType base      = index - baseIndex;
     indexType elem      = base / BITS_PER_ELEMENT;
     indexType posi      = base % BITS_PER_ELEMENT;
 
+    hashBvNode* result = nodeArr[0];
+
     // this should be the 99% case :  when there is only one node in the structure
-    if ((result = nodeArr[0]) && result->baseIndex == baseIndex)
+    if ((result != nullptr) && (result->baseIndex == baseIndex))
```
Comment (sandreenko) [inline src/jit/hashbv.cpp]: I thought that maybe it would affect performance but in my experiment like C++ generated the same code.

Comment (mikedn) [reply src/jit/hashbv.cpp]: Luckily the native compiler is much better at optimizing than the JIT :grin:

File: src/jit/gentree.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/gentree.cpp]: I think I'd rather remove the comments. For one line ifdefs like these they just to the noise.

Comment (tannergooding) [reply src/jit/gentree.cpp]: > For one line ifdefs like these they just to the noise.

I agree. The coding conventions also only indicate it is needed at the end of a long or nested `#if/#ifdef`: https://github.com/dotnet/coreclr/blob/c2baf04cd2c2211334949ba12df2e49fd9109728/Documentation/coding-guidelines/clr-jit-coding-conventions.md#75-commenting-ifdefs

**Comment by tannergooding** [comment]
Are there any of the "off by default" warnings included by `/Wall` that would be useful to track enabling in the future?

File: src/jit/gentree.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/jit/gentree.cpp]: I also prefer to leave them off for small regions. As @mikedn points out, they just clutter things up and don't really add any clarity.

**Comment by mikedn** [comment]
> Are there any of the "off by default" warnings included by /Wall that would be useful to track enabling in the future?

No idea, I didn't set out to improve the JIT code analysis story, just trying to make it on par with the internal build. Besides, `/W4` is something you'd want in any serious project. `/Wall` is extremely noisy, typically it can produce thousands of warnings.

File: src/jit/gentree.cpp (hunk 1 above)
Comment (mikedn) [reply src/jit/gentree.cpp]: OK, will remove.

**Comment by mikedn** [comment]
> /Wall is extremely noisy, typically it can produce thousands of warnings

Noisy as in locking up Visual Studio. The last number of warnings I saw before locking up was like 15000...

**Comment by mjsabby** [comment]
@sandreenko Did you intend to not squash and merge? If not, could you please squash and merge.

**Comment by sandreenko** [comment]
> @sandreenko Did you intend to not squash and merge?

I have "create a merge commit" by default, so I usually use that.

>If not, could you please squash and merge.

Why do you prefer "Squash and merge"? I do not see a reason to revert it and then merge again with squash, do you?

**Comment by mjsabby** [comment]
@sandreenko This is more for your future changes than reverting this PR. 

Take a look at this

https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/contributing.md#merging-pull-requests-for-contributors-with-write-access

You should change your "create a merge commit" to "squash and merge" unless you have a real reason to do so.

**Pull request merged**